chain/chain/src/flat_storage_resharder.rs

```diff
@@ -0,0 +1,949 @@
+    const BATCH_SIZE: usize = 10_000;
+
+    let (parent_shard, status) = get_parent_shard_and_status(&resharder);
+
+    info!(target: "resharding", ?parent_shard, "flat storage shard split task: starting key-values copy");
+
+    // Prepare the store object for commits and the iterator over parent's flat storage.
+    let flat_store = resharder.runtime.store().flat_store();
+    let mut iter = flat_store.iter(parent_shard);
+
+    loop {
+        let mut store_update = flat_store.store_update();
+
+        // Process a `BATCH_SIZE` worth of key value pairs.
+        let mut iter_exhausted = false;
+        for _ in 0..BATCH_SIZE {
+            match iter.next() {
+                Some(Ok(kv)) => {
+                    if let Err(err) = shard_split_handle_key_value(kv, &mut store_update, &status) {
+                        error!(target: "resharding", ?err, "failed to handle flat storage key");
+                        return false;
+                    }
+                }
+                Some(Err(err)) => {
+                    error!(target: "resharding", ?err, "failed to read flat storage value from parent shard");
+                    return false;
+                }
+                None => {
+                    iter_exhausted = true;
+                }
+            }
+        }
+
+        // Make a pause to commit and check if the routine should stop.
+        if let Err(err) = store_update.commit() {
+            error!(target: "resharding", ?err, "failed to commit store update");
+            return false;
+        }
+
+        // TODO(Trisfald): metrics and logs
+
+        // If `iter`` is exhausted we can exit after the store commit.
+        if iter_exhausted {
+            break;
+        }
+        if controller.is_interrupted() {
+            return false;
+        }
+    }
+    true
+}
+
+/// Handles the inheritance of a key-value pair from parent shard to children shards.
+fn shard_split_handle_key_value(
+    kv: (Vec<u8>, FlatStateValue),
+    store_update: &mut FlatStoreUpdateAdapter,
+    status: &SplittingParentStatus,
+) -> Result<(), std::io::Error> {
+    let (key, value) = kv;
+    if key.is_empty() {
+        return Err(std::io::Error::new(
+            std::io::ErrorKind::InvalidData,
+            "flat storage key is empty",
+        ));
+    }
+    let key_column_prefix = key[0];
+    let SplittingParentStatus { left_child_shard, right_child_shard, shard_layout, .. } = status;
+
+    // Copies a key value pair to the correct child by matching the account id to the new shard.
+    let copy_kv_to_child =
+        |account_id_parser: fn(&[u8]) -> Result<AccountId, _>| -> Result<(), std::io::Error> {
+            // Derive the shard uid for this account in the new shard layout.
+            let account_id = account_id_parser(&key)?;
+            let new_shard_id = account_id_to_shard_id(&account_id, shard_layout);
+            let new_shard_uid = ShardUId::from_shard_id_and_layout(new_shard_id, &shard_layout);
```

Comment:
What about yellow and red keys? (The ones without account id). 
For the early MVP there is no need to actually handle them but it also shouldn't crash.

Reply:
Ah I see now, you only call this method for green accounts.